pkg/objstore/backoff.go

```diff
@@ -0,0 +1,248 @@
+package objstore
+
+import (
+	"context"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type backoffBucket struct {
+	Bucket
+	runutil.Backoff
+	logger        log.Logger
+	ignoreObjects []string
+
+	retries *prometheus.CounterVec
+}
+
+func NewBucketWithBackoff(bkt Bucket, backoff runutil.Backoff, logger log.Logger, reg prometheus.Registerer, ignoreObjects ...string) InstrumentedBucket {
+	b := &backoffBucket{
+		Bucket:        bkt,
+		Backoff:       backoff,
+		logger:        logger,
+		ignoreObjects: ignoreObjects,
+		retries: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_bucket_backoff_retries_total",
+			Help:        "Total number of retried operations against a bucket.",
+			ConstLabels: prometheus.Labels{"bucket": bkt.Name()},
+		}, []string{"operation"}),
+	}
+
+	for _, op := range []string{
+		OpGet,
+		OpGetRange,
+		OpExists,
+		OpUpload,
+		OpDelete,
+	} {
+		b.retries.WithLabelValues(op)
+	}
+	return b
+}
+
+func (b *backoffBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	const op = OpGet
+
+	r, err := b.Bucket.Get(ctx, name)
+	// Don't retry in case of 404 or if it's an object worth ignoring.
+	if err == nil || b.Bucket.IsObjNotFoundErr(err) || b.isIgnoreObject(name) {
+		return r, err
+	}
+
+	for retry := 1; retry <= b.Backoff.Max(); retry++ {
+		select {
+		case <-ctx.Done():
+			return r, err
+		case <-b.Backoff.Done():
+			return r, err
+		case <-time.After(b.Backoff.Duration(retry)):
+			// Retry again after backoff
+		}
+
+		r, err = b.Bucket.Get(ctx, name)
+		// Don't retry in case of 404 or if it's an object worth ignoring.
+		if err == nil || b.Bucket.IsObjNotFoundErr(err) || b.isIgnoreObject(name) {
+			return r, err
+		}
+
+		level.Warn(b.logger).Log("msg", "get failed", "name", name, "retryAttempt", retry, "error", err)
```

Comment:
I don't think it should be this wrapper responsability to log. What this wrapper could do is wrapping the returned error with the number of retries attempted, but in my opinion it should be the caller responsability to log it (if required).